src/gui/editors/PianoRoll.cpp

```diff
@@ -3413,43 +3487,56 @@ void PianoRoll::paintEvent(QPaintEvent * pe )
 		};
 
 		// -- Begin ghost MIDI clip
-		if( !m_ghostNotes.empty() )
+		if (!m_ghostNotes.empty() && m_ghostVisible.value())
 		{
-			for( const Note *note : m_ghostNotes )
+			// Lambda function that takes a value x and rounds it to the next/previous value evenly divisable by q
+			// a.k.a. quantizing... (unlike integer division this is the same for negative numbers)
+			auto roundDown = [](int x, int q) -> int { return std::floor(static_cast<float>(x) / q) * q; };
+			auto roundUp = [](int x, int q) -> int { return std::ceil(static_cast<float>(x) / q) * q; };
+
+			// Horizontal repeat
+			const bool loop = m_ghostRepeated.value();
+			// Round to whole bars
+			const int ghostLength = roundUp(m_ghostBounds.end - m_ghostBounds.start, TimePos::ticksPerBar());
+			// Offset of the leftmost visible repetition, relative to the original ghost notes
+			const int posOffBegin = loop ? roundDown(m_currentPosition - m_ghostBounds.start, ghostLength) : 0;
+			// Tick to the right of the visible area
+			const int rightmostTick = m_currentPosition + (width() - m_whiteKeyWidth) * TimePos::ticksPerBar() / m_ppb;
+			// Where to stop drawing
+			const int posEnd = loop ? rightmostTick : m_ghostBounds.end.getTicks();
```

Comment:
Why don't you repeat everything completely, i.e., rounding up `rightmostTick` as multiples of `m_ghostBounds.end.getTicks()`?

Reply:
No need to draw/iterate outside visible area